### PR TITLE
fetch: Fix various issues related to `zig fetch --save`

### LINKED ISSUE #18639: `zig fetch --save localpath` generates url+hash instead of just path
### Zig Version

0.12.0-dev.2307+1b8f7e46f

### Steps to Reproduce and Observed Behavior

`lib/build.zig.zon`
```zig
.{
    .name = "lib",
    .version = "0.1.0",
    .paths = .{""},
}
```
`lib/build.zig`
```zig
const std = @import("std");

pub fn build(b: *std.Build) void {
    _ = b.addModule("lib", .{
        .root_source_file = .{ .path = "lib.zig" },
    });
}
```
`lib/lib.zig`
```zig
pub fn get_number() usize {
    return 42;
}
```
_

`app/build.zig.zon`
```zig
.{
    .name = "app",
    .version = "0.1.0",
    .paths = .{""},
}
```
`app/build.zig`
```zig
const std = @import("std");

pub fn build(b: *std.Build) !void {
    const exe = b.addExecutable(.{
        .name = "main",
        .root_source_file = .{ .path = "main.zig" },
        .target = b.host,
    });

    const lib_mod = b.dependency("lib", .{}).module("lib");
    exe.root_module.addImport("lib", lib_mod);
    b.installArtifact(exe);
}
```
`app/main.zig`
```zig
const lib = @import("lib");

pub fn main() !void {
    @import("std").log.debug("{d}", .{lib.get_number()});
}
```
`app` and `lib` share the same parent folder. 

Running:
```
$ cd app
$ zig fetch --save ../lib
```

Results in:
```zig
.{
    .name = "app",
    .version = "0.1.0",
    // new
    .dependencies = .{
        .lib = .{
            .url = "../lib",
            .hash = "1220773b828b56e62e7e70238b4213e99e8b7e4f8ccec8a70e1404bdfe4886bb0adf",
        },
    },
    .paths = .{""},
}
```
Running `zig build` thereafter produces an error:
```
thread 7278714 panic: unable to find module 'lib'
/Users/timfayz/.zig/lib/std/debug.zig:434:22: 0x1021314a3 in panicExtra__anon_14927 (build)
    std.builtin.panic(msg, trace, ret_addr);
                     ^
/Users/timfayz/.zig/lib/std/debug.zig:409:15: 0x102105477 in panic__anon_13700 (build)
    panicExtra(null, null, format, args);
              ^
/Users/timfayz/.zig/lib/std/Build.zig:1730:18: 0x1020c8af7 in module (build)
            panic("unable to find module '{s}'", .{name});
                 ^
/Users/timfayz/dev/test/fetchZonLocally/app/build.zig:4:52: 0x1020c86a7 in build (build)
    const lib_mod = b.dependency("lib", .{}).module("lib");
                                                   ^
/Users/timfayz/.zig/lib/std/Build.zig:1885:43: 0x1020b42fb in runBuild__anon_8021 (build)
        .ErrorUnion => try build_zig.build(b),
                                          ^
/Users/timfayz/.zig/lib/build_runner.zig:319:29: 0x1020af913 in main (build)
        try builder.runBuild(root);
                            ^
/Users/timfayz/.zig/lib/std/start.zig:585:37: 0x1020b4733 in main (build)
            const result = root.main() catch |err| {
                                    ^
???:?:?: 0x10245d0f3 in ??? (???)
???:?:?: 0xc9677fffffffffff in ??? (???)
error: the following build command crashed:
/Users/timfayz/dev/test/fetchZonLocally/app/zig-cache/o/eb64301265b7ff0d532a4f9e65583421/build /Users/timfayz/.zig/zig /Users/timfayz/dev/test/fetchZonLocally/app /Users/timfayz/dev/test/fetchZonLocally/app/zig-cache /Users/timfayz/.cache/zig --seed 0x5be323c9
```

### Expected Behavior

I believe `zig fetch --save ../lib`  should be aware of the filesystem level and generate the following instead:
```zig
    ...
    .dependencies = .{
        .lib = .{
            .path = "../lib",
        },
    },
    ...
}
```
Changing `.url` to `.path` and removing `.hash` fix the error.

### LINKED ISSUE #21690: `zig fetch --save` puts URL in already existing `.path` entry
### Zig Version

0.14.0-dev.1860+2e2927735

### Steps to Reproduce and Observed Behavior

Given a `build.zig.zon` with a path dependency

```zig
.zqlite = .{ .path = "foo" },
```

when you fetch a URL dependency using

    zig fetch --save https://github.com/karlseguin/zqlite.zig/archive/master.tar.gz

it puts the URL in the `path` field, resulting in an incorrect dependency:

```zig
.zqlite = .{ .path = "https://github.com/karlseguin/zqlite.zig/archive/master.tar.gz", },
```

I'm providing a small repository so you can reproduce quickly: https://github.com/dermetfan/zig-issue-fetch-path

### Expected Behavior

I believe it is reasonable for `zig fetch --save` to recognize this and write a valid dependency:

```zig
.zqlite = .{ .url = "https://github.com/karlseguin/zqlite.zig/archive/master.tar.gz", .hash = "<something>" },
```

### PR DESCRIPTION
**Pull request by castholm**
Closes #18639
Closes #21690

- `zig fetch --save file/path` is now an error.
- `--save`ing over an existing `path` entry will now correctly replace it with an `url` entry.
- Fetching by file URL (as in `zig fetch file:///C:/foo/bar.zip`) did not previously work, but does now.

The purpose of `zig fetch` is to copy a package to the global cache to avoid needing to access the network. It is not to manage your `build.zig.zon`. `--save` is provided as a convenience since calculating the hash by hand would be unreasonable, but it does not make sense to `--save` a package fetched by a (CWD-relative) file path since `path` entries in a build.zig.zon are for relative paths inside a package. `path` entries can only be added by manually editing your `build.zig.zon`.

Both `zig fetch` and `build.zig.zon` support fetching by file URL scheme, for example if a package is on an intranet network drive, but this did not previously always work due to a field being uninitialized and due to file URL-encoded absolute paths on Windows containing drive letters requiring extra attention. This PR fixes those bugs and improves validation to ensure file URLs are spec-compliant (based on RFC 8089 and https://url.spec.whatwg.org/).

I've verified the correctness of these changes by running the following commands locally (on Windows):
- `zig fetch --save ..\..\foo\` -> error (can't save when fetching by file path)
- `zig fetch --save ..\..\foo.zip` -> error (can't save when fetching by file path)
- `zig fetch --save file:///foo.zip` -> ok
- `zig fetch --save file:/foo.zip` -> ok
- `zig fetch --save file:///C:/foo.zip` -> ok
- `zig fetch --save file:/C:/foo.zip` -> ok
- `zig fetch --save file://LOCALhost/C:/foo.zip` -> ok
- `zig fetch --save file:foo.zip` -> error (invalid file URL)

Additionally, it turns out that there were a lot of test declarations in `/src` that were unreferenced and untested, some of which were for the package manager internals. I de-bitrotted those tests and added them under a new `test-compiler-internals` test set, which tests everything referenced by `/src/main.zig`.

### CODE REVIEW

File: src/Package/Fetch.zig [hunk 1]
```diff
@@ -65,6 +65,9 @@ oom_flag: bool,
 /// If the resource pointed to by the location is not a Git-repository, this
 /// will be left unchanged.
 latest_commit: ?git.Oid,
+/// If `location` was `path_or_url`, this will be set to indicate
+/// whether the CLI input was identified as a file path or a URL.
+actual_path_or_url_kind: enum { path, url },
```
Comment (castholm) [inline src/Package/Fetch.zig:70]: I just want to note that this does mean that if the user invokes a command like `zig fetch --save ./foo.tar.gz`, the package will first be copied to the global cache before printing the error. This is obviously not perfect; ideally, the command should detect that the user passed a file path in combination with `--save` and exit before doing any work. However, if we want a maximally consistent and correct behavior, each solution comes with its own downsides

- Currently, we test if the `path_or_url` input is a directory we can open, then a file we can open, then finally a URL. Moving this logic to the `cmdFetch()` function before we actually fetch the package would mean we now have open file handles that we need to pass to the `run()` function and correctly close them if any errors occur on the way there (but not if `run()` returns an error since `run()` claims ownership of the handle). This is not trivial and will result in ugly code.
- We could change the order of operations and instead try parsing the input as an URL first. However, an input like `c:foo` (on Windows) or `http:foo` (on POSIX) would be interpreted as a URL despite possibly being a valid file path, so this isn't ideal. Even if we try to rule out values we know are not valid URLs that can be fetched, an input like `http://example.com/example.tar.gz` *could* be a valid file path to the subpath `example.com/example.tar.gz` inside the directory `http:`.
- We could also change the fetch command so that the user explicitly has to specify whether the input is a path or url, as in `zig fetch --path http:foo --save`.

I've thought about this a lot and I'm sure what the best way to handle this is, or if it's even worth putting energy into. I just think edge cases matter, however obscure they might be.

File: build.zig [hunk 2]
```diff
@@ -379,32 +379,6 @@ pub fn build(b: *std.Build) !void {
     const test_target_filters = b.option([]const []const u8, "test-target-filter", "Skip tests whose target triple do not match any filter") orelse &[0][]const u8{};
     const test_slow_targets = b.option(bool, "test-slow-targets", "Enable running module tests for targets that have a slow compiler backend") orelse false;
 
-    const test_cases_options = b.addOptions();
```
Comment (castholm) [inline build.zig:382]: These were unreferenced.

File: build.zig [hunk 3]
```diff
@@ -531,6 +505,21 @@ pub fn build(b: *std.Build) !void {
         .max_rss = 5029889638,
     }));
 
+    test_modules_step.dependOn(tests.addModuleTests(b, .{
+        .test_filters = test_filters,
+        .test_target_filters = test_target_filters,
+        .test_slow_targets = test_slow_targets,
+        .root_src = "src/main.zig",
+        .name = "cli-internals",
```
[the remark this answers is not in the document]
Comment (castholm) [reply build.zig:513]: That would work too, so long as it's not misleading since it tests package manager internals and other things not strictly related to compiling code. I had trouble coming up with a suitable name in the first place, the main point I wanted to get across was that it tests `src/main.zig`, similar to how `test-std` tests `lib/std/std.zig`.

[the remark this answers is not in the document]
Comment (alexrp) [reply build.zig:513]: I'm mostly suggesting `test-compiler-internals` because we already have `test-cli`:

https://github.com/ziglang/zig/blob/e5d9d3f8a128ccf38b7e2f82c6abe70fb34e9eec/test/tests.zig#L1031-L1032

This new step runs all the unit tests defined within the compiler source code, right? In that case, I think `test-cli-internals` would be misleading given the existence of `test-cli` which narrowly tests the CLI.

[the remark this answers is not in the document]
Comment (castholm) [reply build.zig]: I changed it to `test-compiler-internals`.

**Comment by castholm** [comment]
@andrewrk FYI, I rebased the branch and squashed the last two commits since they were already cherry-picked into #22222 and their previous messages no longer made sense. Your force push ran into a spurious CI failure so the checks needed to be rerun anyway.

File: src/Package/Fetch.zig (hunk 1 above)
Comment (castholm) [reply src/Package/Fetch.zig]: 4 months later and I realize I was seriously overthinking this. Just try to parse the path as a URL in advance (but detect Windows drive letters) and treat it as a path if it's not a valid URL. This is much better. For edge cases like a directory named `http:foo` the user can just pass `./http:foo` instead.

File: src/Package/Fetch.zig [hunk 4]
```diff
@@ -274,14 +274,19 @@ pub const Location = union(enum) {
...
-    /// absolute, in which case it should be treated exactly like a `file://`
-    /// URL, or a directory, in which case it should be treated as an
-    /// already-unpacked directory (but still needs to be copied into the
-    /// global package cache and have inclusion rules applied).
-    path_or_url: []const u8,
+    /// A URL passed via the CLI.
+    /// Indicates that the archive the URL refers to should
+    /// be fetched and copied into the global package cache.
+    /// Recursive Fetch tasks will never use this Location.
+    cli_url: std.Uri,
```
Comment (andrewrk) [inline src/Package/Fetch.zig:281]: src/Package/Fetch.zig shouldn't know about the CLI.

**Review by andrewrk — CHANGES_REQUESTED** [review]

File: src/Package/Fetch.zig (hunk 4 above)
Comment (castholm) [reply src/Package/Fetch.zig:281]: Please clarify, is your objection just to the names having "cli" in them, or that there are fields that are only used to serve the needs of `zig fetch` and thus only used by the CLi in practice? If its the names they can be changed, no problem, but if its the latter I don't really see a good way around it.

Comment (castholm) [reply src/Package/Fetch.zig]: Renamed them to just `url` and `path`, do you think that's better? Any other possible problems were present with the old `url_or_path` as well, all I did was break them out into two distinct fields.

**Comment by alexrp** [comment]
This no longer applies cleanly after the build system configurer/maker split.
